### ISSUE TITLE
irepl to replication resource with all local leaf resources throws errors and creates orphan files

### ISSUE THREAD

**Issue by alanking** [irods/irods#4319]
- [x] master
- [x] 4-2-stable
---
### iRODS Version, OS and Version
4.1.12, 4.2.5

### What did you try to do?
```
$ ilsresc                  
replResc:replication
├── ufs1:unixfilesystem # all 3 resources reside on catalog consumer
├── ufs2:unixfilesystem
└── ufs3:unixfilesystem
rootResc:passthru
└── demoResc:unixfilesystem # resides on catalog provider

$ fallocate -l 3M /tmp/3M.test
$ iput -KR rootResc /tmp/3M.test
$ irepl -R replResc 3M.test
```

### Expected behavior
```
$ ils -L 3M.test
  rods              0 rootResc;demoResc      3145728 2019-03-20.11:01 & 3M.test
    sha2:u9Bc9gl6ybH4nqKdJULBt7Z+5GhIOTiV9ankP6H2IeU=    generic    /var/lib/irods/Vault/home/rods/3M.test
  rods              1 replResc;ufs1      3145728 2019-03-20.11:01 & 3M.test
    sha2:u9Bc9gl6ybH4nqKdJULBt7Z+5GhIOTiV9ankP6H2IeU=    generic    /tmp/irods/ufs1vault/home/rods/3M.test
  rods              2 replResc;ufs2      3145728 2019-03-20.11:01 & 3M.test
    sha2:u9Bc9gl6ybH4nqKdJULBt7Z+5GhIOTiV9ankP6H2IeU=    generic    /tmp/irods/ufs2vault/home/rods/3M.test
  rods              3 replResc;ufs3      3145728 2019-03-20.11:01 & 3M.test
    sha2:u9Bc9gl6ybH4nqKdJULBt7Z+5GhIOTiV9ankP6H2IeU=    generic    /tmp/irods/ufs3vault/home/rods/3M.test
```
No errors

### Observed behavior (including steps to reproduce, if applicable)
When running commands above on the catalog provider, all replicas appear correctly on disk and in the catalog. However, errors appear in the log and 0-length orphan files appear on disk:
```
$ ls -l /tmp/irods/ufs1vault/home/rods/3M.test
-rw------- 1 irods irods 3145728 Mar 20 11:01 /tmp/irods/ufs1vault/home/rods/3M.test
$ ls -l /tmp/irods/ufs2vault/home/rods/3M.test
-rw------- 1 irods irods 3145728 Mar 20 11:01 /tmp/irods/ufs2vault/home/rods/3M.test
$ ls -l /tmp/irods/ufs3vault/home/rods/3M.test
-rw------- 1 irods irods 3145728 Mar 20 11:01 /tmp/irods/ufs3vault/home/rods/3M.test
$ ls -l /tmp/irods/ufs2vault/orphan/home/rods/
total 0
-rw------- 1 irods irods 0 Mar 20 11:01 /tmp/irods/ufs2vault/orphan/home/rods/3M.test.490084634
$ ls -l /tmp/irods/ufs3vault/orphan/home/rods/
total 0
-rw------- 1 irods irods 0 Mar 20 11:01 /tmp/irods/ufs3vault/orphan/home/rods/3M.test.4153252307
```
From the catalog consumer log:
```
Mar 20 11:01:30 pid:441 remote addresses: 172.22.0.4 ERROR: sameHostPartialCopy: copy error for 4294944296 status = -23000 SYS_BAD_FILE_DESCRIPTOR
Mar 20 11:01:30 pid:441 remote addresses: 172.22.0.4 ERROR: [-] /tmp/tmpc3zudH/server/api/src/rsFileCreate.cpp:196:int _rsFileCreate(rsComm_t *, fileCreateInp_t *, rodsServerHost_t *, fileCreateOut_t **) :  status [UNIX_FILE_RMDIR_ERR]  errno [Not a directory] -- message [Rmdir error for "/tmp/irods/ufs2vault/home/rods/3M.test", errno = "Not a directory", status = -521020.]
        [-]     /tmp/tmpc3zudH/server/drivers/src/fileDriver.cpp:352:irods::error fileRmdir(rsComm_t *, irods::first_class_object_ptr) :  status [UNIX_FILE_RMDIR_ERR]  errno [Not a directory] -- message [Rmdir error for "/tmp/irods/ufs2vault/home/rods/3M.test", errno = "Not a directory", status = -521020.]
                [-]     /tmp/tmpc3zudH/plugins/resources/replication/librepl.cpp:1009:irods::error repl_file_rmdir(irods::plugin_context &) :  status [UNIX_FILE_RMDIR_ERR]  errno [Not a directory] -- message [Rmdir error for "/tmp/irods/ufs2vault/home/rods/3M.test", errno = "Not a directory", status = -521020.]
                        [-]     /tmp/tmpc3zudH/plugins/resources/unixfilesystem/libunixfilesystem.cpp:1020:irods::error unix_file_rmdir(irods::plugin_context &) :  status [UNIX_FILE_RMDIR_ERR]  errno [Not a directory] -- message [Rmdir error for "/tmp/irods/ufs2vault/home/rods/3M.test", errno = "Not a directory", status = -521020.]

Mar 20 11:01:30 pid:441 remote addresses: 172.22.0.4 ERROR: [-] /tmp/tmpc3zudH/server/api/src/rsFileCreate.cpp:210:int _rsFileCreate(rsComm_t *, fileCreateInp_t *, rodsServerHost_t *, fileCreateOut_t **) :  status [UNIX_FILE_CREATE_ERR]  errno [File exists] -- message [create error for "/tmp/irods/ufs2vault/home/rods/3M.test", errno = "File exists".]
        [-]     /tmp/tmpc3zudH/server/drivers/src/fileDriver.cpp:38:irods::error fileCreate(rsComm_t *, irods::first_class_object_ptr) :  status [UNIX_FILE_CREATE_ERR]  errno [File exists] -- message [create error for "/tmp/irods/ufs2vault/home/rods/3M.test", errno = "File exists".]
                [-]     /tmp/tmpc3zudH/plugins/resources/replication/librepl.cpp:638:irods::error repl_file_create(irods::plugin_context &) :  status [UNIX_FILE_CREATE_ERR]  errno [File exists] -- message [create error for "/tmp/irods/ufs2vault/home/rods/3M.test", errno = "File exists".]
                        [-]     /tmp/tmpc3zudH/plugins/resources/unixfilesystem/libunixfilesystem.cpp:625:irods::error unix_file_create(irods::plugin_context &) :  status [UNIX_FILE_CREATE_ERR]  errno [File exists] -- message [create error for "/tmp/irods/ufs2vault/home/rods/3M.test", errno = "File exists".]

Mar 20 11:01:30 pid:441 remote addresses: 172.22.0.4 ERROR: sameHostPartialCopy: copy error for 4294944296 status = -23000 SYS_BAD_FILE_DESCRIPTOR
Mar 20 11:01:30 pid:441 remote addresses: 172.22.0.4 ERROR: [-] /tmp/tmpc3zudH/server/api/src/rsFileCreate.cpp:196:int _rsFileCreate(rsComm_t *, fileCreateInp_t *, rodsServerHost_t *, fileCreateOut_t **) :  status [UNIX_FILE_RMDIR_ERR]  errno [Not a directory] -- message [Rmdir error for "/tmp/irods/ufs3vault/home/rods/3M.test", errno = "Not a directory", status = -521020.]
        [-]     /tmp/tmpc3zudH/server/drivers/src/fileDriver.cpp:352:irods::error fileRmdir(rsComm_t *, irods::first_class_object_ptr) :  status [UNIX_FILE_RMDIR_ERR]  errno [Not a directory] -- message [Rmdir error for "/tmp/irods/ufs3vault/home/rods/3M.test", errno = "Not a directory", status = -521020.]
                [-]     /tmp/tmpc3zudH/plugins/resources/replication/librepl.cpp:1009:irods::error repl_file_rmdir(irods::plugin_context &) :  status [UNIX_FILE_RMDIR_ERR]  errno [Not a directory] -- message [Rmdir error for "/tmp/irods/ufs3vault/home/rods/3M.test", errno = "Not a directory", status = -521020.]
                        [-]     /tmp/tmpc3zudH/plugins/resources/unixfilesystem/libunixfilesystem.cpp:1020:irods::error unix_file_rmdir(irods::plugin_context &) :  status [UNIX_FILE_RMDIR_ERR]  errno [Not a directory] -- message [Rmdir error for "/tmp/irods/ufs3vault/home/rods/3M.test", errno = "Not a directory", status = -521020.]

Mar 20 11:01:30 pid:441 remote addresses: 172.22.0.4 ERROR: [-] /tmp/tmpc3zudH/server/api/src/rsFileCreate.cpp:210:int _rsFileCreate(rsComm_t *, fileCreateInp_t *, rodsServerHost_t *, fileCreateOut_t **) :  status [UNIX_FILE_CREATE_ERR]  errno [File exists] -- message [create error for "/tmp/irods/ufs3vault/home/rods/3M.test", errno = "File exists".]
        [-]     /tmp/tmpc3zudH/server/drivers/src/fileDriver.cpp:38:irods::error fileCreate(rsComm_t *, irods::first_class_object_ptr) :  status [UNIX_FILE_CREATE_ERR]  errno [File exists] -- message [create error for "/tmp/irods/ufs3vault/home/rods/3M.test", errno = "File exists".]
                [-]     /tmp/tmpc3zudH/plugins/resources/replication/librepl.cpp:638:irods::error repl_file_create(irods::plugin_context &) :  status [UNIX_FILE_CREATE_ERR]  errno [File exists] -- message [create error for "/tmp/irods/ufs3vault/home/rods/3M.test", errno = "File exists".]
                        [-]     /tmp/tmpc3zudH/plugins/resources/unixfilesystem/libunixfilesystem.cpp:625:irods::error unix_file_create(irods::plugin_context &) :  status [UNIX_FILE_CREATE_ERR]  errno [File exists] -- message [create error for "/tmp/irods/ufs3vault/home/rods/3M.test", errno = "File exists".]
```
From the catalog provider log:
```
Mar 20 11:01:30 pid:3355 NOTICE: remoteDataCopy: rcDataCopy failed
Mar 20 11:01:30 pid:3355 NOTICE: remoteFileCreate: rcFileCreate failed for /tmp/irods/ufs2vault/home/rods/3M.test
Mar 20 11:01:30 pid:3355 NOTICE: remoteDataCopy: rcDataCopy failed
Mar 20 11:01:30 pid:3355 NOTICE: remoteFileCreate: rcFileCreate failed for /tmp/irods/ufs3vault/home/rods/3M.test
```

**Comment by kript** [comment]
Will this also affect 4.1.12 and does it apply to resources that are contained under random leaves (i.e. root -> replication -> random -> random -> resource(s) )?

Asking because we have 300TB to migrate and were looking at using `irsync`...

**Comment by kript** [comment]
Ah, actually, re-reading, does this only apply if the resources are all local to the iES?

**Comment by alanking** [comment]
Right, the issue only manifests itself when running `irepl` as a client connected to the catalog provider on a data object also resident on the catalog provider *to* the replication resource. There's some nuance which I failed to capture above and is somewhat difficult to explain concisely... Basically, in order to reproduce the issue, the "selected resource" (aka the first resource to receive a replica) must replicate to a resource which resides on the same host machine provided the `irepl` is initially sending data *over the network*. If it's ALL local, everything is fine. As shown above, all the data arrives correctly in place, but it clutters up the log and creates a bunch of 0-length orphan files. I have not tested this in a federated environment.

Also, I have updated the issue to indicate that this affects 4.1.12.
